src/silicone/multiple_infillers/split_collection_with_remainder_emissions.py

```diff
@@ -0,0 +1,196 @@
+        unit_equivs = all_units["unit"].map(_remove_equivs).drop_duplicates()
+        if len(unit_equivs) != 1 and desired_unit_eqiv != "Mt CO2/yr":
+            raise ValueError(
+                "The variables in this dataframe have units that cannot "
+                "easily be converted to make them consistent."
+            )
+        return relevant_df, desired_unit
+
+    def infill_components(
+        self,
+        aggregate,
+        components,
+        remainder,
+        to_infill_df,
+        cruncher_class=QuantileRollingWindows,
+        use_ar4_data=False,
+        **kwargs
+    ):
+        """
+        Derive the relationship between the composite variables and their sum, then use
+        this to deconstruct the sum.
+
+        Parameters
+        ----------
+        aggregate : str
+            The variable for which we want to calculate timeseries (e.g.
+            ``"Emissions|CO2"``).
+
+        components : list[str]
+            The variables to be infilled directly by the cruncher. (e.g.
+            ``["Emissions|CO2|AFOLU", "Emissions|CO2|Energy"]``).
+
+        remainder : str
+            The variable which will absorb any difference between the aggregate and
+            component emissions. This may be positive or negative. E.g.
+            ``"Emissions|CO2|Industry"``
+
+        to_infill_df : :obj:`pyam.IamDataFrame`
+            The dataframe that already contains the ``aggregate`` variable, but needs
+            the ``components`` to be infilled.
+
+        cruncher_class : :class:
+            The cruncher used to perform the infilling. By default this will be
+            QuantileRollingWindows.
+
+        use_ar4_data : bool
+            Only used if the variables have different units. If true, we convert all
+            values to Mt CO2 equivalent using the IPCC AR4
+            GWP100 data, otherwise (by default) we use the GWP100 data from AR5.
+
+        **kwargs :
+            An optional dictionary of instructions handed to the cruncher.
+
+        Returns
+        -------
+        :obj:`pyam.IamDataFrame`
+            The infilled data resulting from the calculation.
+        """
+        assert (
+            aggregate in to_infill_df.variables().values
+        ), "The database to infill does not have the aggregate variable"
+        assert all(
+            y not in [remainder] + components for y in to_infill_df.variables().values
+        ), "The database to infill already has some component variables"
+        assert len(to_infill_df.data.columns) == len(self._db.data.columns) and all(
+            to_infill_df.data.columns == self._db.data.columns
+        ), (
+            "The database and to_infill_db fed into this have inconsistent columns, "
+            "which will prevent adding the data together properly."
+        )
+        to_infill_df = to_infill_df.filter(variable=aggregate)
+        to_infill_ag_units = to_infill_df.variables(True)["unit"].values[0]
```

Comment:
do we need to check there's only one unit here?

Reply:
can do